ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -1687,44 +1686,38 @@ private InstallSnapshotReplyProto notifyStateMachineToInstallSnapshot(
         LOG.info("{}: notifyInstallSnapshot: nextIndex is {} but the leader's first available index is {}.",
             getMemberId(), state.getLog().getNextIndex(), firstAvailableLogIndex);
-        try {
-          stateMachine.followerEvent().notifyInstallSnapshotFromLeader(roleInfoProto, firstAvailableLogTermIndex)
-              .whenComplete((reply, exception) -> {
-                if (exception != null) {
-                  LOG.warn("{}: Failed to notify StateMachine to InstallSnapshot. Exception: {}",
-                      getMemberId(), exception.getMessage());
-                  inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogIndex, 0);
-                  return;
-                }
 
-                if (reply != null) {
-                  LOG.info("{}: StateMachine successfully installed snapshot index {}. Reloading the StateMachine.",
-                      getMemberId(), reply.getIndex());
-                  stateMachine.pause();
-                  state.updateInstalledSnapshotIndex(reply);
-                  state.reloadStateMachine(reply.getIndex());
-                  installedSnapshotIndex.set(reply.getIndex());
-                } else {
-                  isSnapshotNull.set(true);
-                  if (LOG.isDebugEnabled()) {
-                    LOG.debug("{}: StateMachine could not install snapshot as it is not available", this);
-                  }
+        // If notifyInstallSnapshot future is done asynchronously, the main thread will go through the downside part.
+        // As the time consumed by statemachine is uncontrollable(e.g. the RocksDB checkpoint could be constantly
+        // increasing, the transmission will always exceed one boundary), we expect that once snapshot installed,
+        // follower could work ASAP. For the rest of time, server can respond snapshot installation in progress.
+
+        // There is another appendLog thread trying appending entries, which returns inconsistency entries with
+        // nextIndex and commitIndex to the leader when install snapshot in progress. The nextIndex is updated when
+        // state.reloadStateMachine. We shall keep this index upgrade synchronized with main thread, or else leader
+        // could get this follower's latest nextIndex from appendEntries, it will not send any notify install snapshot
+        // request, not receive SNAPSHOT_INSTALLED answer and then get stuck in the loop of sending appendEntries and
+        // receiving inconsistency entries.
+        stateMachine.followerEvent().notifyInstallSnapshotFromLeader(roleInfoProto, firstAvailableLogTermIndex)
+            .whenComplete((reply, exception) -> {
+              if (exception != null) {
+                LOG.error("{}: Failed to notify StateMachine to InstallSnapshot. Exception: {}",
+                    getMemberId(), exception.getMessage());
+                inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogIndex, 0);
```

Comment:
seems it is better to use `inProgressInstallSnapshotRequest.set(0)` if an exception is thrown, no need to compare!

Reply:
It should not have the case when an exception happens, the inProgressInstallSnapshotRequest is not the value of ```firstAvailableLogIndex```. On the other hand, this is somehow protection that if the corner case happens, there  
 must be something wrong, and this will leave the follower keeping the state of IN_PROGRESS and alert the maintainer.